File: blind_75_leetcode/python/s0003_longest_substring_without_repeating_characters.py
import unittest


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left = result = 0
        n = len(s)
        for (right, c) in enumerate(s):
            if left + result >= n:
                return result
            if c in s[left:right]:
                result = max(result, right - left)
                while s[left] != s[right]:
                    left += 1
                left += 1
        result = max(result, n - left)
        return result


class Test0003(unittest.TestCase):
    def helper(self, s):
        solution = Solution()
        return solution.lengthOfLongestSubstring(s)

    def test01(self):
        self.assertEqual(self.helper("abcabcbb"), 3)

    def test02(self):
        self.assertEqual(self.helper("bbbbb"), 1)

    def test03(self):
        self.assertEqual(self.helper("pwwkew"), 3)

    def test04(self):
        self.assertEqual(self.helper(" "), 1)

    def test05(self):
        self.assertEqual(self.helper("abcb"), 3)

File: blind_75_leetcode/src/solution/s0003_longest_substring_without_repeating_characters.rs
#![allow(dead_code)]
pub struct Solution {}

impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        let (mut left, mut result) = (0, 0);
        let s: Vec<_> = s.chars().collect();
        let mut set = std::collections::HashSet::new();
        for (right, &c) in s.iter().enumerate() {
            if result + left >= s.len() {
                return result as i32;
            }
            if set.contains(&c) {
                result = result.max(right - left);
                while s[left] != c {
                    set.remove(&s[left]);
                    left += 1;
                }
                left += 1;
            } else {
                set.insert(c);
            }
        }
        result.max(s.len() - left) as i32
    }
}

#[cfg(test)]
mod tests_3 {
    use super::*;

    #[test]
    fn it_works() {
        assert_eq!(
            Solution::length_of_longest_substring("abcabcbb".to_string()),
            3
        );
        assert_eq!(
            Solution::length_of_longest_substring("bbbbb".to_string()),
            1
        );
        assert_eq!(
            Solution::length_of_longest_substring("pwwkew".to_string()),
            3
        );
        assert_eq!(Solution::length_of_longest_substring("abcb".to_string()), 3);
        assert_eq!(Solution::length_of_longest_substring("".to_string()), 0);
        assert_eq!(Solution::length_of_longest_substring(" ".to_string()), 1);
    }
}
